### ISSUE TITLE
Deadline has no effect when unable to resolve hostname

### ISSUE THREAD

**Issue by photex** [grpc/grpc#8519]
Hi folks,

I realize that the deadline concept may not apply in this case, but no matter how I frame it I need to be able to have clients that fail predictably when it's not possible to resolve the hostname of the service.

I'm not seeing any immediate way to configure a timeout for this. Should I avoid providing names to grpc channels and try to resolving them myself?

Cheers!

**Comment by murgatroid99** [comment]
If you make a call with a deadline and the hostname fails to resolve before the call's deadline, the call should fail with a deadline expired status.

There is also a way to wait for the client to connect successfully with a deadline before making any calls, but the interface for that depends on what language you are using.

**Comment by photex** [comment]
If I set the deadline for 2 seconds and make a call to a host I know doesn't resolve; currently my client blocks for a very very very long time. More than a couple minutes. I was convinced that it would hang forever so I let it run overnight. At some point it bailed. But it's certainly not 2 seconds.
This is C++ on OSX and if I pause with a debugger I see that there are two threads. the thread where my code makes the call, and then a thread where grpc is blocking on getaddrinfo.

I just realized that I have timestamps on my logging output so when I get another look I'll note the actual time.

How can figure out what's wrong here?

**Comment by photex** [comment]
Ok, looks like it takes about 35 minutes for DNS resolution to timeout O.o

**Comment by photex** [comment]
If I first use `channel->WaitForConnected()` with my timeout it works as expected and regardless of that, attempting to make a call will sit and block for the aforementioned amount of time.
So checking for the connection first will prevent the client from locking up at least during initialization. 
I'd like to find out if I'm simply doing something incorrectly though that prevents this mechanism from having the desired effect for a regular call.
There are cases where a connection might have worked but won't be possible for a period of time. I need to be able to handle these situations.

**Comment by photex** [comment]
*head smack*

After way too much time (doing other stuff thankfully) I finally realized that the timeout for some calls was previously using milliseconds for timeouts (how the previous RPC library interpreted, and in an effort to avoid having to start using std::chrono::duration everywhere, I just converted the int provided into seconds. :/
~35 min turns out to be 33.3333 minutes or so when you set the deadline to be 2000 seconds. Everybody has to use std::chrono::duration now instead and so problem solved. Sorry to waste your time!
